### PR TITLE
[CS] Make resolveLocatorToDecl simpler and more robust

### PR DESCRIPTION
**Pull request by hamishknight**
Rather than attempting to pull out a concrete callee reference from the rewritten function expression, this PR changes the logic such that we retrieve the callee from the solution using a callee locator computed before the apply is rewritten, and then pass this callee down through `finishApply` and `coerceCallArguments`.

Resolves SR-11648.

### CODE REVIEW

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -3022,9 +3041,17 @@ namespace {
...
-                           cs.getConstraintLocator(expr)));
+      auto *calleeLoc = CalleeLocators[expr];
+      assert(calleeLoc);
+
+      // Resolve the callee for the application if we have one.
+      ConcreteDeclRef callee;
+      if (auto overload = solution.getOverloadChoiceIfAvailable(calleeLoc)) {
+        auto *decl = overload->choice.getDeclOrNull();
+        callee = resolveConcreteDeclRef(decl, calleeLoc);
+      }
```
Comment (hamishknight) [inline lib/Sema/CSApply.cpp]: I'm not entirely happy with the fact that we cache the resolved concrete decl refs on `ExprRewriter`, as it means we can't just use `resolveLocatorToDecl` here. I can't immediately think of a better alternative though. Ideally we would cache them on the `Solution`, but that would require a `const_cast` which I believe would be UB if we ever had a `const Solution`.

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Comment by hamishknight** [comment]
@swift-ci please smoke test

File: lib/Sema/CSApply.cpp [hunk 2]
```diff
@@ -5364,7 +5196,7 @@ static bool hasCurriedSelf(ConstraintSystem &cs, ConcreteDeclRef callee,
 }
 
 Expr *ExprRewriter::coerceCallArguments(
-    Expr *arg, AnyFunctionType *funcType,
+    Expr *arg, AnyFunctionType *funcType, ConcreteDeclRef callee,
```
Comment (xedin) [inline lib/Sema/CSApply.cpp]: Nit: Could you please follow the same style here and put new argument in a separate line? At least to make it look consistent...

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (xedin) [reply lib/Sema/CSApply.cpp]: I think that this is _okay_ but might be good to add a comment for posterity to outline the reason why we have so many `resolveConcreteDeclRef`s :)

**Review by xedin — APPROVED** [review]
Looks great, thank you!

File: lib/Sema/CSApply.cpp (hunk 2 above)
Comment (hamishknight) [reply lib/Sema/CSApply.cpp]: Sure thing! I'll also follow clang-format's suggestion and move the params up to the opening paren.

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (hamishknight) [reply lib/Sema/CSApply.cpp]: Thanks for the review! I've gone ahead and added a few comments, do they look good?

Comment (xedin) [reply lib/Sema/CSApply.cpp]: Looks good, thank you!

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Pull request merged**